optional/abstract_optional.py

```diff
@@ -1,47 +1,60 @@
 from abc import abstractmethod
-
 from .compatible_abc import CompatibleABC
+from typing import TypeVar, Generic, Callable, Union, NoReturn
+
+
+T = TypeVar('T')
 
 
-class AbstractOptional(CompatibleABC):
+class AbstractOptional(CompatibleABC, Generic[T]):
 
     @abstractmethod
     def is_empty(self):
+        # type: () -> bool
         pass
 
     @abstractmethod
     def get(self):
+        # type: () -> Union[T, NoReturn]
         pass
 
     @abstractmethod
     def get_or_default(self, default_value):
+        # type: (T) -> T
         pass
 
     @abstractmethod
     def get_or_raise(self, exception):
+        # type: (type) -> Union[T, NoReturn]
```

Comment:
`Union` means it can return either thing.
`NoReturn` means it doesnt return, in this case its an exception.